src/main/java/com/minecolonies/coremod/entity/ai/citizen/undertaker/EntityAIWorkUndertaker.java

```diff
@@ -0,0 +1,471 @@
+            }
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Attempt to resurrect buried citizen from its citizen data
+     * Randomize to see if resurrection successful and resurrect if need be
+     *
+     * @return the next IAIState
+     */
+    private IAIState tryResurrect()
+    {
+        @Nullable final BuildingGraveyard buildingGraveyard = getOwnBuilding();
+
+        if (buildingGraveyard == null || checkForToolOrWeapon(ToolType.SHOVEL) || buildingGraveyard.getLastGraveData() == null)
+        {
+            return IDLE;
+        }
+
+        worker.getCitizenData().setVisibleStatus(RESURRECT_ICON);
+        worker.getCitizenStatusHandler().setLatestStatus(new TranslationTextComponent("com.minecolonies.coremod.status.resurrecting"));
+        unequip();
+
+        @Nullable final BlockPos gravePos = buildingGraveyard.getGraveToWorkOn();
+
+        // Still moving to the block
+        if (walkToBlock(gravePos, 3))
+        {
+            return getState();
+        }
+
+        final TileEntity entity = world.getTileEntity(gravePos);
+        if (entity instanceof TileEntityGrave)
+        {
+            if (effortCounter < EFFORT_RESURRECT)
+            {
+                worker.swingArm(Hand.MAIN_HAND);
+                effortCounter += getSecondarySkillLevel();
+                return getState();
+            }
+            effortCounter = 0;
+
+            shouldDumpInventory = true;
+            final double chance = getResurrectChance(buildingGraveyard);
+
+            if (chance >= random.nextDouble())
+            {
+                final ICitizenData citizenData = buildingGraveyard.getColony().getCitizenManager().resurrectCivilianData(buildingGraveyard.getLastGraveData().getCitizenDataNBT(), true, world, gravePos);
+                LanguageHandler.sendPlayersMessage(buildingGraveyard.getColony().getImportantMessageEntityPlayers(), "com.minecolonies.coremod.resurrect", citizenData.getName());
+                worker.getCitizenColonyHandler().getColony().setNeedToMourn(false, buildingGraveyard.getLastGraveData().getCitizenName());
+                AdvancementUtils.TriggerAdvancementPlayersForColony(worker.getCitizenColonyHandler().getColony(), playerMP -> AdvancementTriggers.CITIZEN_RESURRECT.trigger(playerMP));
+                buildingGraveyard.setLastGraveData(null);
+                world.setBlockState(gravePos, Blocks.AIR.getDefaultState());
+                return INVENTORY_FULL;
+            }
+        }
+
+        return DIG_GRAVE;
+    }
+
+    /**
+     * Calculate chance of resurrection from multiple factor: Undertaker Skill, Building Level, Research, Mystical Sites in the city
+     *
+     * @param buildingGraveyard
+     * @return the chance of resurrection
+     */
+    private double getResurrectChance(@NotNull final BuildingGraveyard buildingGraveyard)
+    {
+        double chance = buildingGraveyard.getBuildingLevel() * RESURRECT_BUILDING_LVL_WEIGHT +
```

Comment:
We had a long discussion about this on the discord about 2 weeks ago with a few devs. The system will not be switch but just improved by adding way for the player the boost the undertaker chance to resurrect. 

Removing the current system would mean removing 1 state in the AI, 4 research, the reason for the Mana skill for the undertaker, the reason to wander (1 extra state). I do not believe it will confuse players to just add more resurrection change in the future.

We could re-open this discussion on discord again but we will turn in round and round.

Do you think we can leave it as-is for now ?

Reply:
I think its weird if we now give players the current researches, they use them/Get kinda cheaper ressurection this way, and then we remove them, add the totems and make what they got for cheap beforehand more expensive. Its much more accepted when the undertaker just has no resurrection and simply gets added the additional functionality later

Reply:
I don't remember discussing removing any of the current research with the team. Only adding the totem. (but I may be wrong).

See the comment here for a summary of what was discussed with the team: https://github.com/ldtteam/minecolonies/pull/6737#issuecomment-799808444

`Future addition in another PR
Totem BONUS:
If the Undertaker has a totem in his inventory: the resurrection chance and cap are increased by+5%,
With 2 totem the resurrection chance and cap are increased by +7.5%
The use of totem by the Undertaker is locked by a research (with the totem symbol)
Providing a totem to the Undertaker will give an achievement
The totem have a 1% break chance on each attempt of resurrection`

We don't have any mention of removing any current mechanics, just adding some.

Reply:
FI: conclusion of the discussion:

![image](https://user-images.githubusercontent.com/6545998/114300551-28defc80-9ac1-11eb-9024-7cfd4fe0d128.png)

Reply:
well the discussion we were having was about the complete ressurection feature, not just additions to it

Reply:
Let's wait for the inputs of others as we will not agree just you and I. 

In my opinion, yes, the discussion was about the feature, but the conclusion of this discussion, written here from the team inputs, and reviewed by the team, as well as the history readable in discord clearly state something different than what you remember. My feeling, that may be misplaced, is that you are trying to push your own version of the feature while dismissing other's opinion and input, conscientiously or not.

However, I understand that you will not agree with that previous paragraph, so we won't come to a conclusion without external input.